### ISSUE TITLE
Array of strings isn't properly casted as repeated field in big query

### ISSUE THREAD

**Issue by HarenBroog** [confluentinc/kafka-connect-bigquery#96]
**version:**  `confluent-hub install wepay/kafka-connect-bigquery:2.1.0`

Hi!

There is a problem with casting record with given schema:
```json
{
      "name": "my_array",
      "type": {
        "items": "string",
        "type": "array"
      }
}
```

During table creation in big query, connector sets this field schema as `String` 
![image](https://user-images.githubusercontent.com/6953951/113156947-be190000-923a-11eb-9894-42e4125874c3.png). 

Importing rows ends with error:
```
[row index 495]: invalid: Array specified for non-repeated field.
```
I think schema for that field, should be created with "repeated" option in BigQuery.

**Comment by C0urante** [comment]
It looks like this _should_ be covered by existing code:

https://github.com/confluentinc/kafka-connect-bigquery/blob/v2.1.0/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/convert/BigQuerySchemaConverter.java#L165-L167

https://github.com/confluentinc/kafka-connect-bigquery/blob/v2.1.0/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/convert/BigQuerySchemaConverter.java#L223

And we have a test case that covers arrays of floats:

https://github.com/confluentinc/kafka-connect-bigquery/blob/v2.1.0/kcbq-connector/src/test/resources/integration_test_cases/matryoshka-dolls/schema.json#L10-L15

@HarenBroog Could you provide your connector config and Connect worker config (with any sensitive details such as credentials removed)? Once I have those I'll try to reproduce and hopefully we can move forward from there.

**Comment by HarenBroog** [comment]
Yep, I saw tests for this one, thus it's very strange.

Connector:
```json
{
      "name": "big-query-sink-connector-my_topic",
      "config": {
        "connector.class": "com.wepay.kafka.connect.bigquery.BigQuerySinkConnector",
        "tasks.max": "2",
        "topics": "my_topic",
        "sanitizeTopics": true,
        "autoCreateTables": true,
        "autoUpdateSchemas": true,
        "schemaRetriever": "com.wepay.kafka.connect.bigquery.retrieve.IdentitySchemaRetriever",
        "bufferSize": "100000",
        "maxWriteSize": "10000",
        "tableWriteWait": "1000",
        "project": "my_project",
        "defaultDataset": "dataset",
        "keyfile": "/var/keys/big-query-key.json",
        "bigQueryPartitionDecorator": false,
        "bigQueryMessageTimePartitioning": false,
        "timestampPartitionFieldName": "partition_key",
        "errors.deadletterqueue.topic.name": "my_topic",
        "errors.deadletterqueue.topic.replication.factor": 3,
        "allowNewBigQueryFields": true,
        "allowBigQueryRequiredFieldRelaxation": true,
        "allowSchemaUnionization": true,
        "includeKafkaData": true,
      }
    }
```

kafka-connect is set up with config map like:
```yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: kafka-connect-config
  namespace: kafka-connect
data:
  CONNECT_REST_ADVERTISED_HOST_NAME: "connect"
  CONNECT_PLUGIN_PATH: "/usr/share/java,/usr/share/confluent-hub-components"
  BOOTSTRAP_SERVERS: "kafka.servers:9092"
  BASIC_AUTH_CREDENTIALS_SOURCE: "USER_INFO"
  SCHEMA_REGISTRY_URL: https://registry_url.com
  CONNECT_BOOTSTRAP_SERVERS: "kafka.servers:9092"
  CONNECT_REST_PORT: "8083"
  CONNECT_GROUP_ID: "my-group"
  CONNECT_CONFIG_STORAGE_TOPIC: "my-group-configs"
  CONNECT_OFFSET_STORAGE_TOPIC: "my-group-offsets"
  CONNECT_STATUS_STORAGE_TOPIC: "my-group-status"
  CONNECT_REPLICATION_FACTOR: "3"
  CONNECT_CONFIG_STORAGE_REPLICATION_FACTOR: "3"
  CONNECT_OFFSET_STORAGE_REPLICATION_FACTOR: "3"
  CONNECT_STATUS_STORAGE_REPLICATION_FACTOR: "3"
  CONNECT_LOG4J_ROOT_LOGLEVEL: "INFO"
  CONNECT_LOG4J_LOGGERS: "org.reflections=ERROR"
  # CLASSPATH required due to CC-2422
  CLASSPATH: "/usr/share/java/monitoring-interceptors/monitoring-interceptors-6.1.0.jar"
  # Converters
  CONNECT_KEY_CONVERTER: "org.apache.kafka.connect.storage.StringConverter"
  CONNECT_VALUE_CONVERTER: "io.confluent.connect.avro.AvroConverter"
  CONNECT_VALUE_CONVERTER_ENHANCED_AVRO_SCHEMA_SUPPORT: "true"
  CONNECT_VALUE_CONVERTER_SCHEMAS_ENABLE: "true"
  CONNECT_VALUE_CONVERTER_VALUE_SUBJECT_NAME_STRATEGY: "io.confluent.kafka.serializers.subject.RecordNameStrategy"
  CONNECT_VALUE_CONVERTER_SCHEMA_REGISTRY_URL: https://registry_url.com
  CONNECT_VALUE_CONVERTER_BASIC_AUTH_CREDENTIALS_SOURCE: USER_INFO
  # Connect worker
  CONNECT_SECURITY_PROTOCOL: "SASL_SSL"
  CONNECT_SASL_MECHANISM: "PLAIN"
  # Connect producer
  CONNECT_PRODUCER_SECURITY_PROTOCOL: "SASL_SSL"
  CONNECT_PRODUCER_SASL_MECHANISM: "PLAIN"
  CONNECT_PRODUCER_INTERCEPTOR_CLASSES: "io.confluent.monitoring.clients.interceptor.MonitoringProducerInterceptor"
  CONNECT_PRODUCER_CONFLUENT_MONITORING_INTERCEPTOR_SECURITY_PROTOCOL: "SASL_SSL"
  CONNECT_PRODUCER_CONFLUENT_MONITORING_INTERCEPTOR_SASL_MECHANISM: "PLAIN"
  # Connect consumer
  CONNECT_CONSUMER_SECURITY_PROTOCOL: "SASL_SSL"
  CONNECT_CONSUMER_SASL_MECHANISM: "PLAIN"
  CONNECT_CONSUMER_INTERCEPTOR_CLASSES: "io.confluent.monitoring.clients.interceptor.MonitoringConsumerInterceptor"
  CONNECT_CONSUMER_CONFLUENT_MONITORING_INTERCEPTOR_SECURITY_PROTOCOL: "SASL_SSL"
  CONNECT_CONSUMER_CONFLUENT_MONITORING_INTERCEPTOR_SASL_MECHANISM: "PLAIN"
```

Dockerfile:
```dockerfile
FROM confluentinc/cp-server-connect-base:6.1.0

ENV CONNECT_PLUGIN_PATH="/usr/share/java,/usr/share/confluent-hub-components"

RUN confluent-hub install --no-prompt debezium/debezium-connector-postgresql:1.4.0
RUN confluent-hub install --no-prompt wepay/kafka-connect-bigquery:2.1.0
RUN confluent-hub install --no-prompt confluentinc/connect-transforms:latest
```

**Comment by C0urante** [comment]
Hmmm... I don't see anything obviously wrong with your configuration, like a bad converter or no table creation logic.

I also created a test case for this by adding this snippet to the `BigQuerySinkConnectorIT` class:

```java
    List<List<Object>> expectedIssue96Rows = new ArrayList<>();
    expectedIssue96Rows.add(Arrays.asList(
        1L,
        Arrays.asList("v1", "v2")
    ));
    expectedIssue96Rows.add(Arrays.asList(
        2L,
        Arrays.asList("v3", "v4")
    ));
    testCases.put("issue96", expectedIssue96Rows);
```

And these files to the `kcbq-connector/src/test/resources/integration_test_cases/issue96` directory:

`schema.json`:
```json
{ "type": "record",
  "name": "issue96",
  "namespace": "com.wepay.kafka.connect.bigquery",
  "fields":
  [
    { "name":  "row", "type":  "int"},
    { "name":  "my_array", "type": {"type":  "array" , "items":  "string"}}
  ]
}
```

`data.json`:
```json
{"row":1, "my_array":["v1", "v2"]}
{"row": 2, "my_array":  ["v3", "v4"]}
```

The test case passed and the table was created with the appropriate array schema.

At this point the best guess I can come up with is that either the table already existed with that schema and the connector didn't actually create it, or (since the `RecordNameStrategy` is in use for the Avro converter) the connector is seeing different schemas for the same table and the first one it saw had a nullable string schema for the `my_array` field and the second one had a string array schema for it.

The first case should be trivial to rule out. The second may be a bit trickier and I'm happy to lend a hand on it, but we'll need a reliable reproduction case first.

**Comment by HarenBroog** [comment]
Thank you for checking that out!

Those both scenarios which you mentioned would be probable, but in terms of 1, I've tested it many times, each time deleting whole table at the beginning. From what I think, the second one is not the case either, as that particular field is mapped exactly the same in each record which has it (some just haven't got it).

To ensure that we will eliminate possibility of problem number 2, I will filter out some records on the topic in order to leave there only one schema.

But... by looking at `schema.json` and `data.json` it seems exactly the same as in my case. Input AVRO schema is the same, and output data produced by connector also matches (it was printed to the console). The problem is that, the BigQuery schema is not willing to accept that data, because schema was instantiated as `String` without repeated property.

Do the integrations tests verify that a BigQuery schema was created with that property? Maybe there was some kind of change in the API, I don't know 🤔 

Anyway, I will try to dig deeper. At worst I can transform that column to JSON, but I would rather to avoid it :) In the terms of complete reproduction case, you meant some example repo with kafka cluster, registry and connect set up?

**Comment by HarenBroog** [comment]
Additionally I found that BQ table is created here:
https://github.com/confluentinc/kafka-connect-bigquery/blob/master/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java#L232

And then is printed to the console like (sensitive fields sanitized):
```
connect | [2021-04-02 06:32:35,018] INFO Attempting to create table `my_dataset`.`my_topic` with schema Schema{fields=[Field{name=my_field, type=STRING, mode=NULLABLE, description=null, policyTags=null}} (com.wepay.kafka.connect.bigquery.SchemaManager)
```

**Comment by C0urante** [comment]
>Do the integrations tests verify that a BigQuery schema was created with that property? Maybe there was some kind of change in the API, I don't know 🤔

They don't verify the schema for the table, but they do send data to that table after creating it, which should cover your case (which involves failing to write to an automatically-created table).

> In the terms of complete reproduction case, you meant some example repo with kafka cluster, registry and connect set up?

The smallest possible reproduction would be useful here; if you can just provide a couple of records that break the connector, that'd be fantastic. We already have tooling for spinning up an embedded Schema Registry cluster so if that's required we can include it as well, but hopefully we can narrow down the problem to the contents and schemas of records that the connector sees and rule out an upstream converter as the source of the problem.

> From what I think, the second one is not the case either, as that particular field is mapped exactly the same in each record which has it (some just haven't got it).

Ahh, this might be the problem. With schema unionization, we set fields to `NULLABLE` if they're in some schemas but not others. I haven't had time to try to add a repro case for this yet but it might be as simple as sending two records to the connector at once, one with the `my_array` field and one without.

**Comment by HarenBroog** [comment]
Hello @C0urante 👋 

I was able to prepare reproduction setup, available here: https://github.com/HarenBroog/cp-all-in-one/tree/big-query-casting-arrays-reproduction (please follow reproduce.md in cp-all-in-one dir).

**Comment by C0urante** [comment]
I've opened a fix PR for this with https://github.com/confluentinc/kafka-connect-bigquery/pull/99.

**Comment by C0urante** [comment]
@HarenBroog we've put out a couple new releases with the fix; can you give either 2.0.3 or 2.1.2 a try?

**Comment by HarenBroog** [comment]
@C0urante unfortunately, no luck :(

```
connect            | [2021-04-08 20:03:30,993] ERROR Task failed with java.lang.NullPointerException error: null (com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor)
connect            | Exception in thread "pool-10-thread-1" java.lang.NullPointerException
connect            | 	at java.base/java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011)
connect            | 	at java.base/java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1006)
connect            | 	at com.wepay.kafka.connect.bigquery.SchemaManager.updateSchema(SchemaManager.java:251)
connect            | 	at com.wepay.kafka.connect.bigquery.write.row.AdaptiveBigQueryWriter.attemptSchemaUpdate(AdaptiveBigQueryWriter.java:159)
connect            | 	at com.wepay.kafka.connect.bigquery.write.row.AdaptiveBigQueryWriter.performWriteRequest(AdaptiveBigQueryWriter.java:110)
connect            | 	at com.wepay.kafka.connect.bigquery.write.row.BigQueryWriter.writeRows(BigQueryWriter.java:118)
connect            | 	at com.wepay.kafka.connect.bigquery.write.batch.TableWriter.run(TableWriter.java:96)
connect            | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
connect            | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
connect            | 	at java.base/java.lang.Thread.run(Thread.java:834)
connect            | [2021-04-08 20:03:37,791] ERROR WorkerSinkTask{id=my-connector-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted. Error: Some write threads encountered unrecoverable errors: java.lang.NullPointerException; See logs for more detail (org.apache.kafka.connect.runtime.WorkerSinkTask)
connect            | com.wepay.kafka.connect.bigquery.exception.BigQueryConnectException: Some write threads encountered unrecoverable errors: java.lang.NullPointerException; See logs for more detail
connect            | 	at com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor.maybeThrowEncounteredErrors(KCBQThreadPoolExecutor.java:108)
connect            | 	at com.wepay.kafka.connect.bigquery.BigQuerySinkTask.put(BigQuerySinkTask.java:233)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:586)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:329)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:232)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:201)
connect            | 	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
connect            | 	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:234)
connect            | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
connect            | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
connect            | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
connect            | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
connect            | 	at java.base/java.lang.Thread.run(Thread.java:834)
connect            | [2021-04-08 20:03:37,798] WARN WorkerSinkTask{id=my-connector-0} Offset commit failed during close (org.apache.kafka.connect.runtime.WorkerSinkTask)
connect            | [2021-04-08 20:03:37,798] ERROR WorkerSinkTask{id=my-connector-0} Commit of offsets threw an unexpected exception for sequence number 1: null (org.apache.kafka.connect.runtime.WorkerSinkTask)
connect            | com.wepay.kafka.connect.bigquery.exception.BigQueryConnectException: Some write threads encountered unrecoverable errors: java.lang.NullPointerException; See logs for more detail
connect            | 	at com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor.maybeThrowEncounteredErrors(KCBQThreadPoolExecutor.java:108)
connect            | 	at com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor.awaitCurrentTasks(KCBQThreadPoolExecutor.java:96)
connect            | 	at com.wepay.kafka.connect.bigquery.BigQuerySinkTask.flush(BigQuerySinkTask.java:160)
connect            | 	at com.wepay.kafka.connect.bigquery.BigQuerySinkTask.preCommit(BigQuerySinkTask.java:176)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.commitOffsets(WorkerSinkTask.java:392)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.closePartitions(WorkerSinkTask.java:644)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:202)
connect            | 	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
connect            | 	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:234)
connect            | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
connect            | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
connect            | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
connect            | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
connect            | 	at java.base/java.lang.Thread.run(Thread.java:834)
connect            | [2021-04-08 20:03:37,799] ERROR WorkerSinkTask{id=my-connector-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted (org.apache.kafka.connect.runtime.WorkerTask)
connect            | org.apache.kafka.connect.errors.ConnectException: Exiting WorkerSinkTask due to unrecoverable exception.
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:614)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:329)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:232)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:201)
connect            | 	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
connect            | 	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:234)
connect            | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
connect            | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
connect            | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
connect            | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
connect            | 	at java.base/java.lang.Thread.run(Thread.java:834)
connect            | Caused by: com.wepay.kafka.connect.bigquery.exception.BigQueryConnectException: Some write threads encountered unrecoverable errors: java.lang.NullPointerException; See logs for more detail
connect            | 	at com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor.maybeThrowEncounteredErrors(KCBQThreadPoolExecutor.java:108)
connect            | 	at com.wepay.kafka.connect.bigquery.BigQuerySinkTask.put(BigQuerySinkTask.java:233)
connect            | 	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:586)
connect            | 	... 10 more
connect            | [2021-04-08 20:03:37,803] INFO [Consumer clientId=connector-consumer-my-connector-0, groupId=connect-my-connector] Revoke previously assigned partitions test-0 (org.apache.kafka.clients.consumer.internals.ConsumerCoordinator)
connect            | [2021-04-08 20:03:37,803] INFO [Consumer clientId=connector-consumer-my-connector-0, groupId=connect-my-connector] Member connector-consumer-my-connector-0-c1c3692a-9bf6-49c5-a3c5-85009f798579 sending LeaveGroup request to coordinator broker:29092 (id: 2147483646 rack: null) due to the consumer is being closed (org.apache.kafka.clients.consumer.internals.AbstractCoordinator)
connect            | [2021-04-08 20:03:37,807] INFO Metrics scheduler closed (org.apache.kafka.common.metrics.Metrics)
connect            | [2021-04-08 20:03:37,807] INFO Closing reporter org.apache.kafka.common.metrics.JmxReporter (org.apache.kafka.common.metrics.Metrics)
connect            | [2021-04-08 20:03:37,808] INFO Metrics reporters closed (org.apache.kafka.common.metrics.Metrics)
connect            | [2021-04-08 20:03:37,811] INFO App info kafka.consumer for connector-consumer-my-connector-0 unregistered (org.apache.kafka.common.utils.AppInfoParser)
connect            | [2021-04-08 20:03:37,812] INFO [Producer clientId=connector-dlq-producer-my-connector-0] Closing the Kafka producer with timeoutMillis = 9223372036854775807 ms. (org.apache.kafka.clients.producer.KafkaProducer)
connect            | [2021-04-08 20:03:37,815] INFO Metrics scheduler closed (org.apache.kafka.common.metrics.Metrics)
connect            | [2021-04-08 20:03:37,815] INFO Closing reporter org.apache.kafka.common.metrics.JmxReporter (org.apache.kafka.common.metrics.Metrics)
connect            | [2021-04-08 20:03:37,815] INFO Metrics reporters closed (org.apache.kafka.common.metrics.Metrics)
connect            | [2021-04-08 20:03:37,815] INFO App info kafka.producer for connector-dlq-producer-my-connector-0 unregistered (org.apache.kafka.common.utils.AppInfoParser)
connect            | [2021-04-08 20:03:38,918] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] Attempt to heartbeat failed since group is rebalancing (org.apache.kafka.clients.consumer.internals.AbstractCoordinator)
connect            | [2021-04-08 20:03:38,918] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] Revoke previously assigned partitions  (org.apache.kafka.clients.consumer.internals.ConsumerCoordinator)
connect            | [2021-04-08 20:03:38,919] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] (Re-)joining group (org.apache.kafka.clients.consumer.internals.AbstractCoordinator)
connect            | [2021-04-08 20:03:38,923] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] Successfully joined group with generation Generation{generationId=26, memberId='connector-consumer-my-connector-1-789b5f76-798d-440b-bf74-dd2501b9861f', protocol='range'} (org.apache.kafka.clients.consumer.internals.AbstractCoordinator)
connect            | [2021-04-08 20:03:38,923] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] Finished assignment for group at generation 26: {connector-consumer-my-connector-1-789b5f76-798d-440b-bf74-dd2501b9861f=Assignment(partitions=[test-0])} (org.apache.kafka.clients.consumer.internals.ConsumerCoordinator)
connect            | [2021-04-08 20:03:38,926] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] Successfully synced group in generation Generation{generationId=26, memberId='connector-consumer-my-connector-1-789b5f76-798d-440b-bf74-dd2501b9861f', protocol='range'} (org.apache.kafka.clients.consumer.internals.AbstractCoordinator)
connect            | [2021-04-08 20:03:38,927] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] Notifying assignor about the new Assignment(partitions=[test-0]) (org.apache.kafka.clients.consumer.internals.ConsumerCoordinator)
connect            | [2021-04-08 20:03:38,927] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] Adding newly assigned partitions: test-0 (org.apache.kafka.clients.consumer.internals.ConsumerCoordinator)
connect            | [2021-04-08 20:03:38,928] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] Found no committed offset for partition test-0 (org.apache.kafka.clients.consumer.internals.ConsumerCoordinator)
connect            | [2021-04-08 20:03:38,933] INFO [Consumer clientId=connector-consumer-my-connector-1, groupId=connect-my-connector] Resetting offset for partition test-0 to position FetchPosition{offset=0, offsetEpoch=Optional.empty, currentLeader=LeaderAndEpoch{leader=Optional[broker:29092 (id: 1 rack: null)], epoch=0}}. (org.apache.kafka.clients.consumer.internals.SubscriptionState)
connect            | [2021-04-08 20:03:39,831] INFO Attempting to create table `my_dataset`.`test` with schema Schema{fields=[Field{name=normal_field, type=STRING, mode=REQUIRED, description=null, policyTags=null}]} (com.wepay.kafka.connect.bigquery.SchemaManager)
```
and schema is created like (using reproduction setup from above):
![image](https://user-images.githubusercontent.com/6953951/114089934-0d080a80-98b7-11eb-9955-173f5af0f804.png)

**Comment by C0urante** [comment]
Hmm... interesting. I've created an integration test case for this that closely approximates the reproduction case and it's able to successfully create and write to the table (in `BigQuerySinkConnectorIT.java`):

```java
  @Test
  public void testSchemaEvolution() throws Exception {
    final int tasksMax = 1;
    final String connectorName = "bigquery-connector";
    final String topic = TEST_CASE_PREFIX + "schema-evolution";

    TableClearer.clearTables(newBigQuery(), dataset(), suffixedAndSanitizedTable(topic));

    JsonConverter converter = new JsonConverter();
    converter.configure(Collections.emptyMap(), false);

    Map<String, String> connectorConfig = connectorProps(tasksMax);
    connectorConfig.put(ConnectorConfig.VALUE_CONVERTER_CLASS_CONFIG, JsonConverter.class.getName());
    connectorConfig.put(SinkConnectorConfig.TOPICS_CONFIG, topic);
    connectorConfig.put(BigQuerySinkConfig.ALLOW_SCHEMA_UNIONIZATION_CONFIG, "true");
    connect.configureConnector(connectorName, connectorConfig);

    connect.kafka().createTopic(topic);

    waitForConnectorToStart(connectorName, tasksMax);

    Schema schema1 = SchemaBuilder.struct()
        .field("row", Schema.INT64_SCHEMA);
    Schema schema2 = SchemaBuilder.struct()
        .field("row", Schema.INT64_SCHEMA)
        .field("new_array", SchemaBuilder.array(Schema.STRING_SCHEMA));

    Struct value1 = new Struct(schema1);
    value1.put("row", 1L);

    Struct value2 = new Struct(schema2);
    value2.put("row", 2L);
    value2.put("new_array", Arrays.asList("v1", "v2"));

    String serializedValue1 = new String(converter.fromConnectData(topic, schema1, value1));
    String serializedValue2 = new String(converter.fromConnectData(topic, schema2, value2));

    connect.kafka().produce(topic, serializedValue1);
    connect.kafka().produce(topic, serializedValue2);

    waitForCommittedRecords(
        "bigquery-connector", Collections.singleton(topic), 2, tasksMax, TimeUnit.MINUTES.toMillis(60));

    List<List<Object>> expectedRows = new ArrayList<>();
    expectedRows.add(Arrays.asList(1L, Arrays.asList()));
    expectedRows.add(Arrays.asList(2L, Arrays.asList("v1", "v2")));

    verify("schema-evolution", expectedRows);
  }
```

I'll try with the reproduction setup you've linked on v2.1.2 and see if we can find the difference between that and the integration test case. Appreciate your patience with this!

**Comment by C0urante** [comment]
Okay, I found the difference. In my integration test case the schema with the new array field came after the schema without it; in the reproduction case you provided, the ordering was reversed.

I tweaked my integration test case to match that ordering and was able to reproduce a failure (though not the `NullPointerException` in the stack trace you've provided). I've addressed this with https://github.com/confluentinc/kafka-connect-bigquery/pull/101, and also verified by building locally, adding the Confluent Hub archive for that snapshot into the reproduction case you've provided, and running my locally-patched connector with the instructions for your repro case. The fix appears to work for both the integration test and the reproduction case; the table is created with the correct schema and populated with the expected data.

I suspect the NPE in your stack trace may be caused by deleting and then recreating a table in quick succession; when this happens, things can get a bit hairy. If this surfaces reliably, let me know; I wasn't able to trigger it at all when running through the Dockerized repro case.
